cirq/linalg/transformations.py

```diff
@@ -304,3 +317,152 @@ def partial_trace(tensor: np.ndarray,
+    `keep_indices`, the method will fall back to return `default`. If `default`
+    is not provided, the method will fail and raise `ValueError`.
+
+    Args:
+        wavefunction: A wavefunction to express over a qubit subset.
+        keep_indices: Which indices to express the wavefunction on.
+        default: Determines the fallback behavior when `wavefunction` doesn't
+            have a pure state factorization. If the factored state is not pure
+            and `default` is not set, a ValueError is raised. If default is set
+            to a value, that value is returned.
+        atol: The minimum tolerance for comparing the output state's coherence
+            measure to 1.
+
+    Returns:
+        The wavefunction expressed over the desired subset of qubits.
+
+    Raises:
+        ValueError: if the wavefunction is not of the correct shape or the
+        indices are not a valid subset of the input wavefunction's indices, or
+        the result of factoring is not a pure state.
+    """
+
+    if not np.log2(wavefunction.size).is_integer():
+        raise ValueError("Input wavefunction of size {} does not represent a "
+                         "state over qubits.".format(wavefunction.size))
+
+    n_qubits = int(np.log2(wavefunction.size))
+    if wavefunction.shape == (wavefunction.size,):
+        return_shape = -1
+        wavefunction = wavefunction.reshape((2,) * n_qubits)
+    elif wavefunction.shape == (2,) * n_qubits:
+        return_shape = (2,) * len(keep_indices)
+    else:
+        raise ValueError(
+            "Input wavefunction must be shaped like (2 ** n,) or (2, 2, ...)")
+
+    keep_dims = 1 << len(keep_indices)
+    if not np.isclose(np.linalg.norm(wavefunction), 1):
+        raise ValueError("Input state must be normalized.")
+    if len(set(keep_indices)) != len(keep_indices):
+        raise ValueError(
+            "keep_indices were {} but must be unique.".format(keep_indices))
+    if any([ind >= n_qubits for ind in keep_indices]):
+        raise ValueError(
+            "keep_indices {} are an invalid subset of the input wavefunction.")
+
+    keep_dims = 1 << len(keep_indices)
+    other_qubits = sorted(set(range(n_qubits)) - set(keep_indices))
+    candidates = [
+        wavefunction[
+            predicates.slice_for_qubits_equal_to(other_qubits, k)
+        ].reshape(keep_dims)
+        for k in range(1 << len(other_qubits))
+    ]
+    # The coherence measure is computed using unnormalized candidates.
+    best_candidate = max(candidates, key=lambda c: np.linalg.norm(c, 2))
+    best_candidate = best_candidate / np.linalg.norm(best_candidate)
+    left = np.conj(best_candidate.reshape((keep_dims,))).T
+    coherence_measure = sum([
+            abs(np.dot(left, c.reshape((keep_dims,)))) ** 2
+            for c in candidates
+        ])
+
+    if approx_eq(coherence_measure, 1, atol=atol):
+        return best_candidate.reshape(return_shape)
```

Comment:
Multiply the result by `np.exp(2j * np.pi * random.random())` to ensure users don't accidentally rely on the global phase of what we're returning, which will change unpredictably if we update the method.

Reply:
cool, I like that idea.